x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/bulk/TransportBulkShardOperationsAction.java

```diff
@@ -52,27 +56,72 @@ public TransportBulkShardOperationsAction(
     @Override
     protected WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> shardOperationOnPrimary(
             final BulkShardOperationsRequest request, final IndexShard primary) throws Exception {
-        final Translog.Location location = applyTranslogOperations(request, primary, Engine.Operation.Origin.PRIMARY);
-        return new WritePrimaryResult<>(request, new BulkShardOperationsResponse(), location, null, primary, logger);
+        return shardOperationOnPrimary(request.shardId(), request.getOperations(), primary, logger);
+    }
+
+    static WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> shardOperationOnPrimary(
+            final ShardId shardId,
+            final Translog.Operation[] sourceOperations,
+            final IndexShard primary,
+            final Logger logger) throws IOException {
+        final Translog.Operation[] targetOperations = Arrays.stream(sourceOperations).map(operation -> {
+            final Translog.Operation operationWithPrimaryTerm;
+            switch (operation.opType()) {
+                case INDEX:
+                    final Translog.Index index = (Translog.Index) operation;
+                    operationWithPrimaryTerm = new Translog.Index(
+                            index.type(),
+                            index.id(),
+                            index.seqNo(),
+                            primary.getPrimaryTerm(),
+                            index.version(),
+                            index.versionType(),
+                            index.source().toBytesRef().bytes,
```

Comment:
I think this is dangerous - it's correct now since the way we stream guarantee that the bytes ref is offseted with 0 and has a length equal to the bytes array but this isn't guaranteed by the API. I think we should honor the API or, if this turns out difficult assert at the very least.